thetis/inversion_tools.py

```diff
@@ -487,26 +488,29 @@ def construct_evaluator(self):
         # Construct timeseries interpolator
         self.station_interpolators = []
         self.local_station_index = []
-        for i in range(self.fs_points_0d.dof_dset.size):
-            # loop over local DOFs and match coordinates to observations
-            # NOTE this must be done manually as VertexOnlyMesh reorders points
-            x_mesh, y_mesh = mesh0d.coordinates.dat.data[i, :]
-            xy_diff = xy - numpy.array([x_mesh, y_mesh])
-            xy_dist = numpy.sqrt(xy_diff[:, 0]**2 + xy_diff[:, 1]**2)
-            j = numpy.argmin(xy_dist)
-            self.local_station_index.append(j)
-
-            x, y = xy[j, :]
-            t = self.observation_time[j]
-            v = self.observation_values[j]
-            x_mesh, y_mesh = mesh0d.coordinates.dat.data[i, :]
-
-            msg = 'bad station location ' \
-                f'{j} {i} {x} {x_mesh} {y} {y_mesh} {x-x_mesh} {y-y_mesh}'
-            assert numpy.allclose([x, y], [x_mesh, y_mesh]), msg
-            # create temporal interpolator
-            ip = interp1d(t, v, **interp_kw)
-            self.station_interpolators.append(ip)
+        if len(mesh0d.coordinates.dat.data[:]) > 0:
```

Comment:
In parallel, you can have mesh partitions which don't contain any of the Station locations `xy`, so the corresponding processor will hang because `mesh0d` (L474) doesn't have any coordinates (and therefore size). Adding a check that the length is greater than 0 allows any processors without stations in their mesh partition to pass.

Reply:
That's a little odd - do you mean it's hanging in evaluating `self.fs_points_0d.dof_dset.size`? I would have thought that should just evaluate to 0 in that case, and then it should automatically skip the loop. Weird, but anyway....

Reply:
I've found that if I add this print statement anywhere after the creation of `mesh0d`:

`print(len(mesh0d.coordinates.dat.data[:]))`

It will also advance (without needing to add the `if` statement I've added here). Otherwise it hangs once you get to the loop. I can do a `print_output('Here...')` statement up to the `for` loop but not afterwards. I can't do `print('Here...')` statements at any point once it's reached `construct_evaluator` without either 'fix'. Pretty strange.

I've checked this on both my HP and Dell machine and a colleagues - so not unique to a Firedrake install or machine as far as I can tell.

Reply:
Ah ok, so I suspect what's happening is that whenever the `.dat` of that `mesh.coordinates` is first accessed, there is some initialisation code that needs to happens _collectively_ (i.e. on all processes). Without the if statement, on a process where `self.fs_points_0d.dof_dset.size` is zero, the .dat never gets accessed, and so the other processes will be waiting forever for that process to participate. But having the .dat being accessed in the `if` statement apparently is already sufficient to ensure it is being accessed on all processes. So that seems a perfectly good solution.